### ISSUE TITLE
Add `follow` option

### ISSUE THREAD

**Issue by be-rock** [sharkdp/bat#457]
`less` allows you to start it with `less +F file.log` and it will by default tail and follow the file.  `bat` is great for files containing more lines than can be displayed on the terminal screen in that it can automatically page like `less` and can then tail/follow using `F` but if the file I'm intending to tail/follow only has say 10 lines to start with and will be written to after my application starts up, I would like the ability to by default start `bat` in follow mode such as: `bat +F file.log`

**Comment by sharkdp** [comment]
Thank you for the feedback.

Sounds like a reasonable request to me, but I don't really know how to properly implement this. Maybe someone could research how `tail -f`, `less +F`, etc. make this work. Do they watch the file after it has been completely read? Or do they somehow keep the filehandle open (I can't imaging that this works..)?

**Comment by Jarred-Sumner** [comment]
TLDR, guessing:
- read last N characters of the file
- use one of the filesystem watcher libraries to get notified whenever the file is modified
- read last N characters of file, print characters between old last N and new last N

might have to disable some features in order to make this work though :(, e.g:
- how would you do line numbers if you're not reading the entire file?
- would syntax highlighting work well if the syntax is potentially invalid? (as a result of not reading the entire file)

but you could use a heuristic like, if the file is above a certain size, don't bother with syntax highlighting / line numbers when in follow mode. lots of text editors do this (VSCode, Sublime Text, etc)

---

this Server Fault post from 2010 says stuff: https://serverfault.com/questions/161981/can-tail-slow-down-log-writing-speed-on-linux-ext3

but it doesn't point to any code so I don't really trust that answer entirely.

if you look at the [manual for tail](https://www.gnu.org/software/coreutils/manual/html_node/tail-invocation.html) and then find-in-page for "--follow":
> There are two ways to specify how you’d like to track files with this option, but that difference is noticeable only when a followed file is removed or renamed. If you’d like to continue to track the end of a growing file even after it has been unlinked, use --follow=descriptor. This is the default behavior, but it is not useful if you’re tracking a log file that may be rotated (removed or renamed, then reopened). In that case, use --follow=name to track the named file, perhaps by reopening it periodically to see if it has been removed and recreated by some other program. Note that the inotify-based implementation handles this case without the need for any periodic reopening.

> No matter which method you use, if the tracked file is determined to have shrunk, tail prints a message saying the file has been truncated and resumes tracking the end of the file from the newly-determined endpoint.
>
> When a file is removed, tail’s behavior depends on whether it is following the name or the descriptor. When following by name, tail can detect that a file has been removed and gives a message to that effect, and if --retry has been specified it will continue checking periodically to see if the file reappears. When following a descriptor, tail does not detect that the file has been unlinked or renamed and issues no message; even though the file may no longer be accessible via its original name, it may still be growing.
> 
> The option values ‘descriptor’ and ‘name’ may be specified only with the long form of the option, not with -f.
> 
> The -f option is ignored if no file operand is specified and standard input is a FIFO or a pipe. Likewise, the -f option has no effect for any operand specified as ‘-’, when standard input is a FIFO or a pipe.
>
> With kernel inotify support, output is triggered by file changes and is generally very prompt. Otherwise, tail sleeps for one second between checks— use --sleep-interval=n to change that default—which can make the output appear slightly less responsive or bursty. When using tail without inotify support, you can make it more responsive by using a sub-second sleep interval, e.g., via an alias like this:

it'd be better to dig through `tail`'s source code to see the implementation...but my knowledge of C is pretty crap

**Comment by sharkdp** [comment]
@Jarred-Sumner Thank you for doing the research! Unfortunately, that sounds rather complicated.

I just found out that we can just do the following:
```
tail -f /tmp/foo.txt | bat --paging=never
```

This work just as expected. Including line numbers and syntax highlighting (when explicitly specified via `-l`).

**Comment by be-rock** [comment]
This works exactly as expected, thank you.  Although in a pinch, that is a bit much to type so the following approach meets the objective:

`function batf() { tail -f "$1" | bat --paging=never; }`

And then calling `batf /tmp/foo.txt' (or an alias to it) works

Would it be possible to have the above represented via a `bat -F` argument?

**Comment by sharkdp** [comment]
Given how much pain the `--pager`/`--paging` options have caused (are still causing), I'd rather not implement this and ask users to use the compositional pipe version (possibly through a shell function, as you suggest). I have already added this to the list in #448.

**Comment by be-rock** [comment]
I am ok with this, thanks again for all your great work!

**Comment by jankeesvw** [comment]
I have an improved version of the `batf` function above, this one also does coloring:

```bash
function batf() {
  local filename="$1"
  local extension="${filename##*.}"
  tail -f "$filename" | bat -l "$extension" --paging=never
}
```
